core/src/main/java/com/graphhopper/routing/ProfileResolver.java

```diff
@@ -18,118 +18,125 @@
             if (edgeBased == null && matchingProfiles.size() == 2 &&
-                    match1.getWeighting().getName().equals(match2.getWeighting().getName()) &&
-                    match1.getWeighting().getFlagEncoder().toString().equals(match2.getWeighting().getFlagEncoder().toString()) &&
-                    match1.isEdgeBased() != match2.isEdgeBased()) {
-                return match1.isEdgeBased() ? match1 : match2;
+                    match1.getWeighting().equals(match2.getWeighting()) &&
+                    match1.getVehicle().equals(match2.getVehicle()) &&
+                    match1.isTurnCosts() != match2.isTurnCosts()) {
+                return match1.isTurnCosts() ? match1 : match2;
             }
-            // special case: error if multiple edge-based matches. to differentiate between these it will be required
-            // to explicitly set the profile parameter.
-            if (numMatchingEdgeBased > 1 && numMatchingEdgeBased == matchingProfiles.size()) {
-                throw new IllegalArgumentException("There are multiple edge-based CH profiles matching your request. You need to" +
-                        " specify the profile you want to use explicitly, see here: https://github.com/graphhopper/graphhopper/pull/1934.");
-            } else {
-                throw new IllegalArgumentException("There are multiple CH profiles matching your request. Use the `weighting`,`vehicle`,`edge_based` and/or `u_turn_costs` parameters to be more specific." +
-                        "\nYou can also try disabling CH altogether using " + Parameters.CH.DISABLE + "=true" +
-                        "\nrequested:  " + getCHRequestAsString(hintsMap, edgeBased, uTurnCosts) + "\nmatched:   " + matchingProfiles + "\navailable: " + chProfiles);
+            throw new IllegalArgumentException("There are multiple CH profiles matching your request. Use the `weighting`,`vehicle`,`edge_based` and/or `u_turn_costs` parameters to be more specific or better use the `profile` parameter to explicitly choose a profile." +
+                    "\nYou can also try disabling CH altogether using " + Parameters.CH.DISABLE + "=true" +
+                    "\nrequested:  " + getCHRequestAsString(hintsMap, edgeBased, uTurnCosts) + "\nmatched:   " + chProfilesAsString(matchingProfiles) + "\navailable: " + chProfilesAsString(chProfiles));
 
-            }
         }
     }
 
-    public LMProfile selectLMProfile(List<LMProfile> lmProfiles, HintsMap hintsMap) {
-        List<LMProfile> matchingProfiles = new ArrayList<>();
-        for (LMProfile p : lmProfiles) {
+    protected boolean chProfileMatchesHints(ProfileConfig p, HintsMap hintsMap) {
+        Boolean edgeBased = getEdgeBased(hintsMap);
+        Integer uTurnCosts = getUTurnCosts(hintsMap);
+        return (edgeBased == null || p.isTurnCosts() == edgeBased) &&
+                (uTurnCosts == null || uTurnCosts.equals(getUTurnCosts(p.getHints()))) &&
```

Comment:
Since we are now selecting a `CHProfileConfig` instead of a `CHProfile` we can deal with the case of multiple edge-based profiles with different u_turn_costs again (c.f. #1949). This should also help to get rid of `Weighting#getFlagEncoder` in the long run.